### PR TITLE
Fixbug #3494 No to_python (by-value) converter found for C++ t…

### PR DESCRIPTION
**Pull request by errord**
fixbug #3494 , register shared_ptr Blob Dtype

### CODE REVIEW

**Comment by errord** [comment]
what is label JL?

**Comment by crowsonkb** [comment]
It would be nice if this were merged. I'm having to merge it manually by hand to build on OS X, since Homebrew is packaging boost 1.60.0.

**Comment by shelhamer** [comment]
@crowsonkb Sorry, I can't reproduce this in `master` @ fe0f441. Can you give more details on your setup? I have compiled and tested pycaffe with boost 1.60.0_1 on OS X 10.11.3 without trouble.

**Comment by sbaechler** [comment]
I have the same issue, but when I add this code, Python crashes when I run it.

    Loaded network /Users/simon/py-faster-rcnn/data/faster_rcnn_models/VGG16_faster_rcnn_final.caffemodel
    F0228 14:54:50.696149 1985011712 roi_pooling_layer.cu:91] Check failed: error == cudaSuccess (8 vs. 0)  invalid device function

I'm using the demo.py file from https://github.com/rbgirshick/py-faster-rcnn.

Python 2.7.11, OSX El Capitan, Boost 1.6.0, CUDA support.

**Comment by graphaelli** [comment]
I've also run into this.  There is was no error at build time but you can reproduce it by running tests.  Specifcally - `python3 -m unittest test_solver.py`:

 Without this patch:
```
======================================================================
ERROR: test_net_memory (test_solver.TestSolver)
Check that nets survive after the solver is destroyed.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/caffe/python/caffe/test/test_solver.py", line 29, in setUp
    size=self.solver.net.blobs['label'].data.shape)
TypeError: No to_python (by-value) converter found for C++ type: boost::shared_ptr<caffe::Net<float> >

======================================================================
ERROR: test_snapshot (test_solver.TestSolver)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/caffe/python/caffe/test/test_solver.py", line 29, in setUp
    size=self.solver.net.blobs['label'].data.shape)
TypeError: No to_python (by-value) converter found for C++ type: boost::shared_ptr<caffe::Net<float> >

======================================================================
ERROR: test_solve (test_solver.TestSolver)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/caffe/python/caffe/test/test_solver.py", line 29, in setUp
    size=self.solver.net.blobs['label'].data.shape)
TypeError: No to_python (by-value) converter found for C++ type: boost::shared_ptr<caffe::Net<float> >

----------------------------------------------------------------------
Ran 3 tests in 0.026s

FAILED (errors=3)
```

With:
```
.
----------------------------------------------------------------------
Ran 3 tests in 0.048s

OK
```

**Comment by ahundt** [comment]
I think I just reproduced the problem in #3575 on master with #3570 and #3613 applied (the latter two may not be relevant). When I applied #3575 I was able to continue on. Here is the trace:

Traceback (most recent call last):
  File "/Users/athundt/source/eclipse/caffe/fcn-32s/eval-fcn-32s.py", line 17, in <module>
    net.blobs['data'].reshape(1, *in_.shape)
  File "/Users/athundt/source/git/caffe/python/caffe/pycaffe.py", line 28, in _Net_blobs
    return OrderedDict(zip(self._blob_names, self._blobs))
TypeError: No to_python (by-value) converter found for C++ type: boost::shared_ptr<caffe::Blob<float> >

± python --version
Python 2.7.11

**Comment by panovr** [comment]
I hope this commit can be merged, because I also run into this problem, and my environment is Mac OS X 10.11.3, Xcode 7.2, Anaconda python 2.7.11, boost 1.6.0.

======================================================================
ERROR: test_forward_backward (test_net.TestNet)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/ylzhao/Project/caffe/python/caffe/test/test_net.py", line 45, in setUp
    size=self.net.blobs['label'].data.shape)
  File "caffe/pycaffe.py", line 28, in _Net_blobs
    return OrderedDict(zip(self._blob_names, self._blobs))
TypeError: No to_python (by-value) converter found for C++ type: boost::shared_ptr<caffe::Blob<float> >

======================================================================
ERROR: test_inputs_outputs (test_net.TestNet)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/ylzhao/Project/caffe/python/caffe/test/test_net.py", line 45, in setUp
    size=self.net.blobs['label'].data.shape)
  File "caffe/pycaffe.py", line 28, in _Net_blobs
    return OrderedDict(zip(self._blob_names, self._blobs))
TypeError: No to_python (by-value) converter found for C++ type: boost::shared_ptr<caffe::Blob<float> >

======================================================================
ERROR: test_memory (test_net.TestNet)
Check that holding onto blob data beyond the life of a Net is OK
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/ylzhao/Project/caffe/python/caffe/test/test_net.py", line 45, in setUp
    size=self.net.blobs['label'].data.shape)
  File "caffe/pycaffe.py", line 28, in _Net_blobs
    return OrderedDict(zip(self._blob_names, self._blobs))
TypeError: No to_python (by-value) converter found for C++ type: boost::shared_ptr<caffe::Blob<float> >

======================================================================
ERROR: test_save_and_read (test_net.TestNet)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/ylzhao/Project/caffe/python/caffe/test/test_net.py", line 45, in setUp
    size=self.net.blobs['label'].data.shape)
  File "caffe/pycaffe.py", line 28, in _Net_blobs
    return OrderedDict(zip(self._blob_names, self._blobs))
TypeError: No to_python (by-value) converter found for C++ type: boost::shared_ptr<caffe::Blob<float> >

======================================================================
ERROR: test_net_memory (test_solver.TestSolver)
Check that nets survive after the solver is destroyed.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/ylzhao/Project/caffe/python/caffe/test/test_solver.py", line 29, in setUp
    size=self.solver.net.blobs['label'].data.shape)
TypeError: No to_python (by-value) converter found for C++ type: boost::shared_ptr<caffe::Net<float> >

======================================================================
ERROR: test_snapshot (test_solver.TestSolver)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/ylzhao/Project/caffe/python/caffe/test/test_solver.py", line 29, in setUp
    size=self.solver.net.blobs['label'].data.shape)
TypeError: No to_python (by-value) converter found for C++ type: boost::shared_ptr<caffe::Net<float> >

======================================================================
ERROR: test_solve (test_solver.TestSolver)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/ylzhao/Project/caffe/python/caffe/test/test_solver.py", line 29, in setUp
    size=self.solver.net.blobs['label'].data.shape)
TypeError: No to_python (by-value) converter found for C++ type: boost::shared_ptr<caffe::Net<float> >

----------------------------------------------------------------------
Ran 21 tests in 1.499s

**Comment by shelhamer** [comment]
Ok, I have reproduced the problem and the fix. Thanks for the patch!

File: python/caffe/_caffe.cpp [hunk 1]
```diff
@@ -252,6 +252,7 @@ BOOST_PYTHON_MODULE(_caffe) {
     .def("_set_input_arrays", &Net_SetInputArrays,
         bp::with_custodian_and_ward<1, 2, bp::with_custodian_and_ward<1, 3> >())
     .def("save", &Net_Save);
+  bp::register_ptr_to_python<shared_ptr<Net<Dtype> > >();
```
Comment (shelhamer) [inline python/caffe/_caffe.cpp]: @longjon can you comment on this? It seems redundant to me given the `bp::class_` declarations but boost 1.60 complains without it.

**Comment by shelhamer** [comment]
Argh, this causes warnings at least with boost 1.59 on Ubuntu 14.04. This should maybe be conditioned on boost version/platform as gross as adding guards is.

**Pull request merged**